library/alloc/src/collections/btree/node.rs

```diff
@@ -66,68 +66,116 @@ struct LeafNode<K, V> {
-    /// dropping of uninitialized fields. Second, an invariant of internal nodes is that `len + 1`
-    /// edges are initialized and valid, meaning that even when the node is empty (having a
-    /// `len` of 0), there must be one initialized and valid edge. This function does not set up
-    /// such an edge.
-    unsafe fn new() -> Self {
-        InternalNode { data: unsafe { LeafNode::new() }, edges: MaybeUninit::uninit_array() }
+    /// # Safety
+    /// Every item returned by `range` is a valid edge index for the node.
+    unsafe fn correct_childrens_parent_links<R: Iterator<Item = usize>>(&mut self, idx_range: R) {
+        let ptr = NonNull::new(self);
+        for idx in idx_range {
+            let child = unsafe { self.edges.get_unchecked_mut(idx).assume_init_mut().ptr.as_mut() };
+            child.parent = ptr;
+            child.parent_idx = MaybeUninit::new(idx as u16);
+        }
+    }
+}
+
+pub trait NodeTypeTrait<K, V>
+where
+    Self: Sized,
+{
+    type E;
+
+    /// Wrapper for the `correct_childrens_parent_links` method on internal nodes,
+    /// doing nothing on leaf nodes.
+    ///
+    /// # Safety
+    /// Every item returned by `range` is a valid edge index for the node.
+    unsafe fn correct_childrens_parent_links<R: Iterator<Item = usize>>(
+        _node: &mut Node<K, V, Self>,
+        _idx_range: R,
+    ) {
+    }
+}
+
+impl<K, V> NodeTypeTrait<K, V> for marker::Leaf {
+    type E = ();
+}
+
+impl<K, V> NodeTypeTrait<K, V> for marker::Internal {
+    type E = BoxedNode<K, V>;
+
+    unsafe fn correct_childrens_parent_links<R: Iterator<Item = usize>>(
+        node: &mut Node<K, V, Self>,
+        idx_range: R,
+    ) {
+        unsafe { node.correct_childrens_parent_links(idx_range) }
     }
 }
 
-/// A managed, non-null pointer to a node. This is either an owned pointer to
-/// `LeafNode<K, V>` or an owned pointer to `InternalNode<K, V>`.
+// This acts as a type alias that is externally visible,
+// allowing gdb_providers.py to introspect.
+#[repr(transparent)]
+struct InternalNode<K, V> {
+    // gdb_providers.py uses this field name for introspection.
+    _wraps: Node<K, V, marker::Internal>,
+}
+
+/// A managed, non-null pointer to a leaf or internal `Node`.
 ///
 /// However, `BoxedNode` contains no information as to which of the two types
 /// of nodes it actually contains, and, partially due to this lack of information,
 /// has no destructor.
-struct BoxedNode<K, V> {
+pub struct BoxedNode<K, V> {
     ptr: Unique<LeafNode<K, V>>,
 }
 
 impl<K, V> BoxedNode<K, V> {
-    fn from_leaf(node: Box<LeafNode<K, V>>) -> Self {
-        BoxedNode { ptr: Box::into_unique(node) }
-    }
-
-    fn from_internal(node: Box<InternalNode<K, V>>) -> Self {
+    fn from_new<Type>(node: Box<Node<K, V, Type>>) -> BoxedNode<K, V>
+    where
+        Type: NodeTypeTrait<K, V>,
+    {
         BoxedNode { ptr: Box::into_unique(node).cast() }
```

Comment:
I hoped to get rid of this cast, just boxing the address of the `leaf`, but whatever I do, I get runtime errors, reproducible but depending on the exact code, sometimes only in one unit test. So inexplicable I worry that the old and/or the proposed code still has undefined behaviour.

Reply:
That was caused by another stupid mistake.